superset-frontend/src/dashboard/util/isValidChild.ts

```diff
@@ -105,12 +105,36 @@ const parentMaxDepthLookup = {
   [MARKDOWN_TYPE]: {},
 };
 
-export default function isValidChild({ parentType, childType, parentDepth }) {
+type ParentMaxDepthLookup = typeof parentMaxDepthLookup;
+
+interface IsValidChildProps {
+  parentType?: unknown;
+  childType?: unknown;
+  parentDepth?: unknown;
+}
+
+interface ValidChild<
+  P extends keyof ParentMaxDepthLookup,
+  C extends keyof ParentMaxDepthLookup[P],
+  D extends ParentMaxDepthLookup[P][C]
+> {
+  parentType: P;
+  childType: C;
+  parentDepth: D;
+}
+
+export default function isValidChild<
+  P extends keyof ParentMaxDepthLookup = any,
+  C extends keyof ParentMaxDepthLookup[P] = any,
+  D extends ParentMaxDepthLookup[P][C] = any
+>(child: IsValidChildProps): child is ValidChild<P, C, D> {
+  const { parentType, childType, parentDepth } = child;
   if (!parentType || !childType || typeof parentDepth !== 'number') {
```

Comment:
if we have TypeScript, is type check here still necessary?

Reply:
It depends, I aimed to keep the behavior the same, taking in an unknown and returning a known type.

This entire function could theoretically be completely removed, using a combo of keyof and generics can statically check most of this:

```
interface ValidChild<
  P extends keyof ParentMaxDepthLookup,
  C extends keyof ParentMaxDepthLookup[P],
  D extends ParentMaxDepthLookup[P][C]
> {
  parentType: P;
  childType: C;
  parentDepth: D;
}
```

---

as an intermediate step, I could set `parentDepth` to be number and remove this check.

thoughts? :thought_balloon: 

can check all three statically.

Reply:
yes, i feel `parentDepth` should be `number` type.  This intermediate step a little confused me.  Thanks!

Reply:
@graceguo-supercat I removed the generics which were causing confusion.
The request to make `parentDepth` a `number` and remove the `typeof` check causes a cascade affect where two other test suites fail

```
Test Suites: 2 failed, 186 passed, 188 total
Tests:       17 failed, 4 skipped, 1040 passed, 1061 total
Snapshots:   12 passed, 12 total
Time:        67.338s, estimated 78s
```

specifically `getDropPosition` and `isValidChild` both expect the number check.